cmd/admin-handlers.go

```diff
@@ -1058,16 +1076,35 @@ func mustTrace(entry interface{}, trcAll, errOnly bool) bool {
 func (a adminAPIHandlers) TraceHandler(w http.ResponseWriter, r *http.Request) {
 	ctx := newContext(r, w, "HTTPTrace")
 
-	trcAll := r.URL.Query().Get("all") == "true"
-	trcErr := r.URL.Query().Get("err") == "true"
-
 	// Validate request signature.
 	_, adminAPIErr := checkAdminRequestAuth(ctx, r, iampolicy.TraceAdminAction, "")
 	if adminAPIErr != ErrNone {
 		writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(adminAPIErr), r.URL)
 		return
 	}
 
+	trcErr := r.URL.Query().Get("err") == "true"
+	trcS3 := r.URL.Query().Get("s3") == "true"
+	trcInternal := r.URL.Query().Get("internal") == "true"
+	trcStorage := r.URL.Query().Get("storage") == "true"
+
+	// Support deprecated 'all' query
+	if r.URL.Query().Get("all") == "true" {
+		trcS3 = true
+		trcInternal = true
+		trcStorage = true
+	}
+
+	var threshold time.Duration
+	if t := r.URL.Query().Get("threshold"); t != "" {
+		d, err := time.ParseDuration(t)
+		if err != nil {
+			writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(ErrInvalidRequest), r.URL)
+			return
+		}
+		threshold = d
+	}
+
```

Comment:
write it in a function return TraceOptions data structure write ParseTraceOptions(r.Query())